### PR TITLE
Fix issue with in-place modification of FieldInfo

### LINKED ISSUE #2971: Annotated FieldInfo instance is mutated and reused if using type aliases
### Checks

* [x ] I added a descriptive title to this issue
* [ x] I have searched (google, github) for similar issues and couldn't find anything
* [ x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.2
            pydantic compiled: False
                 install path: /tmp/tmp.pnuPT7tTpf/lib/python3.9/site-packages/pydantic
               python version: 3.9.6 (default, Jun 30 2021, 10:22:16)  [GCC 11.1.0]
                     platform: Linux-5.13.0-rc7-x86_64-with-glibc2.33
     optional deps. installed: ['devtools', 'typing-extensions']

```

Working version
```py
from pydantic import BaseModel, constr, Field
from typing import Annotated

class MyModel(BaseModel):
  a: Annotated[str, Field(max_length=3)] = ''
  b: Annotated[str, Field(max_length=3)]

MyModel(a="123", b='123', c='123')
MyModel(a="1234", b='1234', c='1234')  # ValidationError
```

Failing version:
```py
from pydantic import BaseModel, constr, Field
from typing import Annotated

StrField3 = Annotated[str, Field(max_length=3)]

class MyModel(BaseModel):  # ValueError: `Field` default cannot be set in `Annotated` for 'b'
  a: StrField3 = ''
  b: StrField3
```

The `.default` is set to `''` on the FieldInfo instance at [fields.py:404](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/fields.py#L404).

This can be fixed by copying the `FieldInfo` before mutating it and associating it with ModelField.
```
from copy import copy
...
class ModelField(Representation):
     ...
    @staticmethod
    def _get_field_info(
        field_name: str, annotation: Any, value: Any, config: Type['BaseConfig']
    ) -> Tuple[FieldInfo, Any]:
          ....
          if field_info is not None:
               field_info = copy(field_info)
               ....
```

### LINKED ISSUE #3686: `Annotated[..., Field(...)]` cannot be used as a type alias because it is modified in-place
Example code:

```python
from typing import Annotated
from pydantic import BaseModel, Field


FieldAnnotation = Annotated[int, Field(gt=0)]

class Model(BaseModel):
    field1: FieldAnnotation = 1
    field2: FieldAnnotation = 2
```

This gives the error `Field default cannot be set in Annotated for 'field2'`.
This is triggered here:
https://github.com/samuelcolvin/pydantic/blob/6f46a5a1460728868f860840e3f6186aa40403d8/pydantic/fields.py#L451

Because when `field1` is processed, the annotation (which is just an object) gets modified in place:
https://github.com/samuelcolvin/pydantic/blob/6f46a5a1460728868f860840e3f6186aa40403d8/pydantic/fields.py#L454

I actually discovered this because I was (by design) calling `ModelField.infer` twice on the same annotation, but I am reporting it as a type alias issue since I think that is a more common use case / reasonable thing for an end user to do.

### LINKED ISSUE #3702: Adding `list[Annotated[str, Field(min_length=1)]]` before other fields causes unrelated ValueError
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.9.0
            pydantic compiled: True
                 install path: /Users/brianhammond/Desktop/pez/.venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.0 (default, Jan 11 2022, 16:51:17) [Clang 13.0.0 (clang-1300.0.29.30)]
                     platform: macOS-12.1-x86_64-i386-64bit
     optional deps. installed: ['email-validator', 'typing-extensions']
```

```py
from typing import Annotated
from pydantic import Field, BaseModel

NonEmptyStr = Annotated[str, Field(min_length=1)]

class Foo(BaseModel):
    xs: list[NonEmptyStr]
    id: NonEmptyStr

foo_ok = Foo(id="some_id", xs=["hello"])
foo_too_short = Foo(id="", xs=["hello"])
```

Unexpected error:

```
ValueError: `Field` default cannot be set in `Annotated` for 'id'
```

Yet, if the `xs` field is moved after `id`, no problem:

```py
from typing import Annotated
from pydantic import Field, BaseModel

NonEmptyStr = Annotated[str, Field(min_length=1)]


class Foo(BaseModel):
    id: NonEmptyStr
    xs: list[NonEmptyStr]


foo_ok = Foo(id="some_id", xs=["hello"])
foo_too_short = Foo(id="", xs=["hello"])
```

Expected error:
```
pydantic.error_wrappers.ValidationError: 1 validation error for Foo
id
  ensure this value has at least 1 characters (type=value_error.any_str.min_length; limit_value=1)
```

The error occurs also if such a field is defined in another class, but only if defined *earlier*:

```py
from typing import Annotated
from pydantic import Field, BaseModel

NonEmptyStr = Annotated[str, Field(min_length=1)]


class Bar(BaseModel):
    xs: list[NonEmptyStr]


class Foo(BaseModel):
    id: NonEmptyStr


foo_ok = Foo(id="some_id", xs=["hello"])
foo_too_short = Foo(id="", xs=["hello"])
```

Unexpected error:

```
ValueError: `Field` default cannot be set in `Annotated` for 'id'
```

Defined later:

```py
from typing import Annotated
from pydantic import Field, BaseModel

NonEmptyStr = Annotated[str, Field(min_length=1)]


class Foo(BaseModel):
    id: NonEmptyStr


class Bar(BaseModel):
    xs: list[NonEmptyStr]


foo_ok = Foo(id="some_id", xs=["hello"])
foo_too_short = Foo(id="", xs=["hello"])
```

Expected error:

```py
pydantic.error_wrappers.ValidationError: 1 validation error for Foo
id
  ensure this value has at least 1 characters (type=value_error.any_str.min_length; limit_value=1)
```

I'm guessing but it seems like some kind of reference is being held onto between contexts...

Thanks for any help, info, etc.

### LINKED ISSUE #3714: ValueError raised when Nested Discriminated Unions are used twice
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

The Following code is copied from [doc](https://pydantic-docs.helpmanual.io/usage/types/#nested-discriminated-unions) as is except for the last `Model2`:
```py
from typing import Literal, Union

from typing_extensions import Annotated

from pydantic import BaseModel, Field, ValidationError


class BlackCat(BaseModel):
    pet_type: Literal["cat"]
    color: Literal["black"]
    black_name: str


class WhiteCat(BaseModel):
    pet_type: Literal["cat"]
    color: Literal["white"]
    white_name: str


# Can also be written with a custom root type
#
# class Cat(BaseModel):
#   __root__: Annotated[Union[BlackCat, WhiteCat], Field(discriminator='color')]

Cat = Annotated[Union[BlackCat, WhiteCat], Field(discriminator="color")]


class Dog(BaseModel):
    pet_type: Literal["dog"]
    name: str


Pet = Annotated[Union[Cat, Dog], Field(discriminator="pet_type")]


class Model(BaseModel):
    pet: Pet
    n: int

# The above code is copied as is, the following is added
class Model2(BaseModel):
    pet: Pet
    n: int
```

Run the above code, without the last `Model2`, the code runs well, but with `Model2`, the code throws the following error:

```
Traceback (most recent call last):
  File "/Users/stephen/Work/medieco/hub/hub-api/test.py", line 41, in <module>
    class Model2(BaseModel):
  File "pydantic/main.py", line 204, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 488, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 419, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 534, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 599, in pydantic.fields.ModelField._type_analysis
  File "pydantic/fields.py", line 633, in pydantic.fields.ModelField._type_analysis
  File "pydantic/fields.py", line 776, in pydantic.fields.ModelField._create_sub_type
  File "pydantic/fields.py", line 451, in pydantic.fields.ModelField._get_field_info
ValueError: `Field` default cannot be set in `Annotated` for "pet_Annotated[Union[__main__.BlackCat, __main__.WhiteCat], FieldInfo(discriminator='color', extra={})]"
```

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.9.0
            pydantic compiled: True
                 install path: /Users/stephen/work/.venv/lib/python3.9/site-packages/pydantic
               python version: 3.9.0 (default, Dec 20 2020, 11:26:29)  [Clang 12.0.0 (clang-1200.0.32.28)]
                     platform: macOS-12.1-x86_64-i386-64bit
     optional deps. installed: ['dotenv', 'email-validator', 'typing-extensions']
```

### LINKED ISSUE #3804: Using default_factory for Annotated[Union[...], Field(default_factory=...)] prevents using default_factory for collections containing  that type
### Checks

* [ x ] I added a descriptive title to this issue
* [ x ] I have searched (google, github) for similar issues and couldn't find anything
* [ x ] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Trying to run the script below produces this error, tested for both for pydantic 1.8.2 and 1.9.0:

```
Traceback (most recent call last):
  File "/[snip]i/temp.py", line 31, in <module>
    class LetterHolder2(BaseModel):
  File "pydantic/main.py", line 204, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 488, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 419, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 534, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 728, in pydantic.fields.ModelField._type_analysis
  File "pydantic/fields.py", line 776, in pydantic.fields.ModelField._create_sub_type
  File "pydantic/fields.py", line 464, in pydantic.fields.ModelField._get_field_info
  File "pydantic/fields.py", line 211, in pydantic.fields.FieldInfo._validate
ValueError: cannot specify both default and default_factory
```

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`: (Also tested for 1.8.2)
```
             pydantic version: 1.9.0
            pydantic compiled: True
                 install path: /home/USER/.cache/pypoetry/virtualenvs/temp-project-jzM_oSD3-py3.9/lib/python3.9/site-packages/pydantic
               python version: 3.9.10 (main, Jan 15 2022, 18:17:56)  [GCC 9.3.0]
                     platform: Linux-5.10.60.1-microsoft-standard-WSL2-x86_64-with-glibc2.31
     optional deps. installed: ['typing-extensions']
```

```py
from typing import Literal, Annotated, Union, Dict

from pydantic import BaseModel, Field


class A(BaseModel):
    letter_type: Literal["A"] = "A"


class B(BaseModel):
    letter_type: Literal["B"] = "B"


Letter1 = Annotated[Union[A, B], Field(discriminator="letter_type")]


# This definition works
class LetterHolder1(BaseModel):
    mapping_to_letter: Dict[str, Letter1] = Field(default_factory=dict)


def default_constructor() -> A:
    """Default constructor for letters."""
    return A()


Letter2 = Annotated[Union[A, B], Field(discriminator="letter_type", default_factory=default_constructor)]


# This definition does not work
class LetterHolder2(BaseModel):
    mapping_to_letter: Dict[str, Letter2] = Field(default_factory=dict)

# Neither does this one
class LetterHolder3(BaseModel):
    mapping_to_letter: List[Letter2] = Field(default_factory=list)

```

### PR DESCRIPTION
**Pull request by adriangb**
Fixes #2971, fixes #3686, fixes #3702, fixes #3804, fixes #3714

### CODE REVIEW

**Comment by adriangb** [comment]
Need to add a test for #3714, not 100% sure that's fixed

**Comment by adriangb** [comment]
@skruppy I just noticed that you did the same fix in a fork (https://github.com/skruppy/pydantic/commit/6ae42bbe3bad4c075fe2dc885c112778918f4cc4, I guess it's just the obvious solution) and @bcarley said it "breaks other things". Could one of you folks elaborate on this? As far as I can tell no tests break.

File: pydantic/fields.py [hunk 1]
```diff
@@ -446,6 +447,7 @@ def _get_field_info(
                 raise ValueError(f'cannot specify multiple `Annotated` `Field`s for {field_name!r}')
             field_info = next(iter(field_infos), None)
             if field_info is not None:
+                field_info = copy.copy(field_info)
```
Comment (samuelcolvin) [inline pydantic/fields.py:450]: any chance we might need `deepcopy` just to be sure?

File: changes/4067-adriangb.md [hunk 2]
```diff
@@ -0,0 +1 @@
+fix in-place modification of FieldInfo that made it impossible to use PEP 593 type aliases
```
Comment (samuelcolvin) [inline changes/4067-adriangb.md:1]:
```suggestion
Fix in-place modification of `FieldInfo` that caused problems with PEP 593 type aliases
```

File: pydantic/fields.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/fields.py:450]: e.g. perhaps if we had nested annoated?

**Review by samuelcolvin — COMMENTED** [review]
Otherwise I think this looks great.

@BCarley seems to be suggesting this causes errors in tests which doesn't seem to be the case, so I think we're good to go.

**Comment by samuelcolvin** [comment]
please update.

**Comment by samuelcolvin** [comment]
Awesome, thanks so much.

**Pull request merged**